packages/server/src/core/credit_transfer_graph.rs

```diff
@@ -0,0 +1,51 @@
+use petgraph::algo::toposort;
+use petgraph::Graph;
+
+use crate::error::AppError;
+use crate::resources::catalog::Catalog;
+use crate::resources::course::CourseBank;
+
+use super::messages;
+use super::types::CreditOverflow;
+
+fn build_credit_transfer_graph(
+    course_banks: &[CourseBank],
+    credit_overflow_rules: &[CreditOverflow],
+) -> Graph<String, ()> {
+    let mut g = Graph::<String, ()>::new();
+    for course_bank in course_banks {
+        g.add_node(course_bank.name.clone());
+    }
+    for credit_rule in credit_overflow_rules {
+        g.add_edge(
+            // unwrap cannot fail because the credit rules are taken from bank names
+            g.node_indices()
+                .find(|i| g[*i] == credit_rule.from)
+                .unwrap(),
+            g.node_indices().find(|i| g[*i] == credit_rule.to).unwrap(),
+            (),
+        );
+    }
+    g
+}
+
+pub fn find_traversal_order(catalog: &Catalog) -> Vec<CourseBank> {
+    let g = build_credit_transfer_graph(&catalog.course_banks, &catalog.credit_overflows);
+    let order = toposort(&g, None).unwrap(); // unwrap can't fail because if the catalog exists in the database it means it is valid, i.e no cycles.
```

Comment:
This is a tough one.
Though the assumption mentioned in this comment is 100% correct, still we must be prepared to handle even the weirdest of cases where our data gets mistakenly corrupted or we fail to validate a catalog successfully and break this assumption.
I would use `unwrap_or_else` here, giving an arbitrary order to the indices if the toposort fails and report (log) the error.
```rust
let order = toposort(&g, None).unwrap_or_else(|_| {
    log!(
        log::Level::Error,
        "error: corrupted catalog in the database - course banks will be set in an arbitrary order"
    );
    g.node_indices().into_iter().collect::<Vec<_>>()
});
```